devito/ir/support/guards.py

```diff
@@ -136,12 +136,12 @@ def __new__(cls, d, direction, **kwargs):
 
             if d.is_Conditional:
                 v = d.factor
-                # Round `p1` down to the nearest sub-multiple of `v`
-                # NOTE: we use FLOAT(d.factor) to make sure we don't drop negative
-                # values on the floor. E.g., `iteration=time - 1`, `v=2`, then when
-                # `time=0` we want the Mul to evaluate to -1, not to 0, which is
-                # what C's integer division would give us
-                p1 = Mul(((p1 - 1) / FLOAT(v)), v, evaluate=False)
+                # Round `p1 - 1` down to the nearest sub-multiple of `v`
+                # NOTE: we use ABS to make sure we handle negative values properly.
+                # Once `p1 - 1` is negative (e.g. `iteration=time - 1` and `time=0`),
+                # as long as we get a negative number, rather than 0 and even if it's
+                # not `-v`, we're good
+                p1 = (p1 - 1) - abs(p1 - 1) % v
```

Comment:
But that's not a multiple of `v` if `p1-1` is negative, it should be 
`p1 = (p1 - 1) - sign(p1 - 1) * (abs(p1 - 1) % v)`
to always be a multiple of v irrespective of sign.

Reply:
yes, but the comment above explains why we can tolerate it

I agree it's far from ideal, but in practice it always works (unless we start allowing time dimensions in negative ranges...)

The reason I haven't done it like that is because I don't know how to neatly turn `sign` into C code. It would require a non trivial addition to the compiler, basically a further compiler pass intercepting "high order functions", such as `sign`, that don't have a one-to-one matching with routines in clib (math.h)

If we wanted explicit, inline lowering, we could use one of these forms https://stackoverflow.com/questions/1903954/is-there-a-standard-sign-function-signum-sgn-in-c-c
But it feels like it would clutter the generated code...

Unless I'm missing the obvious, I'd prefer to open an issue rather than tackling this now

Reply:
Sign is a defined function we use it for the free surface for example I'm the acoustic propagator there is nothing needed in the compiler

Reply:
Sorry, I forgot to copy-paste what I wrote on Slack here. Here it is:

It crashes in (at least) one case
So the guard that gets generated lets the execution flow pass through at `time = [4, 2, 0]` while the expectation is to let it go through at `time = [4, 2]`